File: A06 Nodes Related/002 Binary Tree/LC116 Populating next right pointers in each node/BFS.java
/*
Time Complexity:  O(n)
Space Complexity:  O(1)

*/ 


/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if (root == null) return root;
        
        Node level_start = root; 
        //start from here
        while(level_start != null){
            Node cur = level_start;
            while(cur != null){
                if(cur.left != null){
                    cur.left.next = cur.right; 
                    if(cur.next != null){
                        cur.right.next = cur.next.left;     //cur.next is current root.next; it was node which connected to current root; 
                    }
                }
                cur = cur.next;
            }
         //end here=> process connection for the whole level
        level_start = level_start.left;
        }
       return root; 
        
    }
}

File: A06 Nodes Related/002 Binary Tree/LC116 Populating next right pointers in each node/BFS.py
###########
Time Complexity:
Space Complexity:


#############
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        curr = root
        while curr and curr.left:
            left = curr.left
            while curr:
                curr.left.next = curr.right
                curr.right.next = curr.next.left if curr.next else None
                curr = curr.next
            curr = left
        return root